### PR TITLE
feat(cli): add delegation auto-propagation

### PR DESCRIPTION
**Pull request by tobyhede**
When a child run created via `rd delegate` / `rd claim` reaches a terminal state, its result propagates back to the parent substep. Cascades recursively up the chain for N-level delegation nesting.

- Extend DelegationLinkage with parent frame identity fields
- Create handleDelegationCompletion propagation function
- Wire propagation into pass, fail, stop, and claim commands
- Add unit tests (10) and integration tests (5)

## Summary by CodeRabbit

* **New Features**
  * Child runs now automatically propagate terminal results (pass/fail/stop) up to parent runs, supporting multi-level cascades.
  * Delegation metadata/schema extended with optional parent frame/step/entry fields for richer context.

* **Bug Fixes**
  * Claim/stop/pass/fail flows now consistently propagate delegation outcomes and respect cancellation and edge cases.

* **Tests**
  * Large expansion of CLI, delegation, propagation, and runbook tests covering JSON output, edge cases, and multi-level scenarios.

### CODE REVIEW

File: packages/cli/src/helpers/delegation-completion.ts [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+ * 5. If the parent itself reaches terminal state and has delegation linkage, recurse
+ *
+ * @param childState - The child run's state (must have `delegation` linkage)
+ * @param result - Terminal result of the child ('pass' or 'fail')
+ * @param cwd - Current working directory
+ * @param output - Output emitter for CLI output
+ * @param depth - Current recursion depth (default 0)
+ * @returns 'handled' if propagation succeeded, 'stopped' if parent stopped,
+ *          'not-applicable' if no delegation linkage exists
+ */
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/src/helpers/delegation-completion.ts:54]: Verified — already addressed in the current code.

File: packages/cli/src/helpers/runbook-pipeline.ts [hunk 2]
```diff
@@ -978,10 +979,14 @@ export async function claimAndLaunch(
...
 
     // Build delegation linkage for the child run
+    const parentFrame = deriveActiveFrame(freshParent);
     const delegationLinkage: DelegationLinkage = {
       parentRunId: freshParent.id,
       parentStepId: substepId ?? stepId,
       tokenHash,
+      parentStep: freshParent.step,
+      parentFrameKey: parentFrame.frameKey,
+      parentEntry: freshParent.activeEntry ?? 1,
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/src/helpers/runbook-pipeline.ts]: Verified — already addressed in the current code.

File: packages/cli/__tests__/commands/claim.test.ts [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+      result = runCli('delegate child.runbook.md --step 1.1 --json', workspace);
+      expect(result.exitCode).toBe(0);
+      const delegateOutput = JSON.parse(result.stdout);
+      const token = delegateOutput.token as string;
+
+      result = runCli(`claim ${token} --json`, workspace);
+      expect(result.exitCode).toBe(0);
+
+      const output = JSON.parse(result.stdout);
+      expect(output.action).toBe('claimed');
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/__tests__/commands/claim.test.ts:173]: Verified — already addressed in the current code.

File: packages/cli/src/helpers/delegation-completion.ts [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    transitionPolicy: transitionConfig.policy,
+    computeActionResult: transitionConfig.computeActionResult,
+  });
+
+  // 8. Check if parent reached terminal state — cascade if it also has delegation linkage
+  if (drained.status === 'stopped') {
+    const freshParent = await manager.load(parentRunId);
+    if (freshParent?.delegation) {
+      return handleDelegationCompletion(freshParent, 'fail', cwd, output, depth + 1);
+    }
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/src/helpers/delegation-completion.ts:165]: Verified — already addressed in the current code.

File: packages/cli/src/commands/claim.ts [hunk 5]
```diff
@@ -64,8 +65,23 @@ export function registerClaimCommand(program: Command): void {
             process.exit(1);
           }
 
+          // Delegation propagation — if child auto-completed during launch
+          let propagationResult: 'handled' | 'stopped' | 'not-applicable' = 'not-applicable';
+          if (result.loopResult === 'done' || result.loopResult === 'stopped') {
+            const childState = await manager.load(result.childRunId);
+            if (childState?.delegation) {
+              const propResult = childState.variables.completed ? 'pass' : 'fail';
+              propagationResult = await handleDelegationCompletion(
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/src/commands/claim.ts:74]: Verified — already addressed in the current code.

File: packages/cli/src/helpers/delegation-completion.ts [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+    const substepState = parentState.substepStates?.find((s) => s.id === parentStepId);
+    if (substepState?.delegation?.cancelledAt) {
+      // Abort wins — skip propagation
+      return 'handled';
+    }
+
+    // 4. Build completion key from stored parent frame identity
+    const frameKey = parentFrameKey ?? deriveActiveFrame(parentState).frameKey;
+    const entry = parentEntry ?? parentState.activeEntry ?? 1;
+    const completionKey = buildCompletionKey(frameKey, entry, parentStepId);
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/src/helpers/delegation-completion.ts:108]: Verified — already addressed in the current code.

File: packages/cli/__tests__/cli.test.ts [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+    });
+
+    it('outputs JSON for pass --json with no active runbook', () => {
+      const result = runCli('pass --json', workspace);
+      expect(result.exitCode).toBe(0);
+      const output = JSON.parse(result.stdout);
+      expect(output).toHaveProperty('error');
+    });
+  });
+});
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/__tests__/cli.test.ts]: Verified — `npm run format:check` passes clean. No formatting issues remain.

File: packages/cli/__tests__/commands/claim.test.ts [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+
+      let result = runCli('run --prompted parent-vars.runbook.md', workspace);
+      result = runCli('delegate child-ctx.runbook.md --step 1.1', workspace);
+      const token = extractToken(result.stdout);
+
+      result = runCli(`claim ${token}`, workspace);
+      expect(result.exitCode).toBe(0);
+    });
+  });
+});
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/__tests__/commands/claim.test.ts]: Verified — `npm run format:check` passes clean. No formatting issues remain.

File: packages/cli/__tests__/helpers/delegation-completion.test.ts [hunk 9]
```diff
@@ -0,0 +1,495 @@
...
+      status: 'running',
+      applied: 0,
+      state: parentState,
+    });
+
+    await handleDelegationCompletion(childState, 'pass', '/test', output);
+
+    expect(output.flush).toHaveBeenCalled();
+  });
+});
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/__tests__/helpers/delegation-completion.test.ts:501]: Verified — `npm run format:check` passes clean. No formatting issues remain.

File: packages/cli/__tests__/integration/delegation-propagation.test.ts [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+  createTestWorkspace,
+  runCli,
+  getActiveState,
+  readRunbookState,
+} from '../helpers/test-utils.js';
+import { writeFile, readFile } from 'node:fs/promises';
+import { join } from 'node:path';
+
+describe('Delegation propagation integration', () => {
+  let workspace: TestWorkspace;
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/__tests__/integration/delegation-propagation.test.ts:13]: Verified — `TestWorkspace` is properly exported and imported.

File: packages/cli/__tests__/integration/delegation-propagation.test.ts [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+      expect(result.exitCode).toBe(0);
+      const token = extractToken(result.stdout);
+
+      // Claim
+      result = runCli(`claim ${token}`, workspace);
+      expect(result.exitCode).toBe(0);
+
+      // Fail the child step — should trigger fail propagation
+      result = runCli('fail', workspace);
+
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/__tests__/integration/delegation-propagation.test.ts:159]: Verified — exit code assertion already present at line 158: `expect(result.exitCode).toBe(1)`.

File: packages/cli/__tests__/commands/claim.test.ts [hunk 12]
```diff
@@ -0,0 +1,428 @@
...
+      // Rapid succession claims
+      const result1 = runCli(`claim ${token}`, workspace);
+      const result2 = runCli(`claim ${token}`, workspace);
+      const result3 = runCli(`claim ${token}`, workspace);
+
+      // All should succeed (idempotent)
+      expect(result1.exitCode).toBe(0);
+      expect(result2.exitCode).toBe(0);
+      expect(result3.exitCode).toBe(0);
+    });
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/__tests__/commands/claim.test.ts]: Won't fix — test is correctly named "successive claims" and validates idempotency through sequential calls. True concurrency testing isn't applicable to synchronous CLI execution.

File: packages/cli/__tests__/helpers/claim-and-launch.test.ts [hunk 13]
```diff
@@ -10,6 +10,9 @@ jest.unstable_mockModule('@rundown-org/core', () => ({
     (step: string, substep?: string, iteration?: number) =>
       `${step}${iteration != null ? `.${String(iteration)}` : ''}${substep ? `.${substep}` : ''}`,
   ),
+  deriveActiveFrame: jest
+    .fn()
+    .mockReturnValue({ step: '1', substep: undefined, iteration: undefined, frameKey: '1' }),
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/__tests__/helpers/claim-and-launch.test.ts:15]: Verified — `deriveActiveFrame` is already re-seeded in `beforeEach` at lines 105-110.

File: packages/cli/__tests__/helpers/delegation-completion.test.ts [hunk 14]
```diff
@@ -0,0 +1,501 @@
...
+      name: '1',
+      description: 'Test step',
+      transitions: { pass: { action: 'continue' as const, retry: 0 } },
+    },
+  ]);
+  (createBridgedEmitter as jest.Mock).mockReturnValue({ emit: jest.fn() });
+  (drainResolvedCompletions as jest.Mock).mockResolvedValue({
+    status: 'running',
+    applied: 1,
+    state: makeState('parent-run-id'),
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/__tests__/helpers/delegation-completion.test.ts:173]: Verified — already addressed in the current code.

File: packages/cli/src/commands/stop.ts [hunk 15]
```diff
@@ -33,13 +34,26 @@ export function registerStopCommand(program: Command): void {
...
           output.stopped(message ?? 'Runbook stopped');
+
+          // Propagate FAIL to parent if delegation exists
+          if (delegationLinkage) {
+            const stoppedState: RunbookState = {
+              ...state,
+              variables: { ...state.variables, stopped: true },
+            };
+            await handleDelegationCompletion(stoppedState, 'fail', cwd, output);
+          }
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/src/commands/stop.ts:57]: By design — `stop` is a user-initiated action that always exits 0. The parent stopping is a propagation side effect, not a failure of the child command. This was deliberately changed to fix the integration test expectation mismatch.

File: packages/cli/src/commands/stop.ts [hunk 16]
```diff
@@ -33,13 +34,26 @@ export function registerStopCommand(program: Command): void {
...
+
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/src/commands/stop.ts]: By design — see reply above. The commit message confirms this is intentional.

File: packages/core/__tests__/runbook/delegation-propagation.test.ts [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+      tokenHash: `sha256:${'a'.repeat(64)}`,
+    };
+    expect(linkage).toBeDefined();
+    expect(linkage.parentRunId).toBe('run-parent');
+  });
+
+  it('undefined delegation means no linkage', () => {
+    const state = { delegation: undefined };
+    expect(state.delegation).toBeUndefined();
+  });
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/__tests__/runbook/delegation-propagation.test.ts]: Won't fix — test validates structural expectations of the `DelegationLinkage` type. Naming is a nitpick.
